// redirect-form-builder/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormBuilder = void 0;
function uniqueId() {
    const dict = "QWERTYUIOPASDFGHJKLZXVBNM1234567890";
    return new Array(8)
        .fill("")
        .map(() => dict[Math.floor(Math.random() * dict.length)])
        .join("");
}
class FormBuilder {
    constructor(url = "", method = "POST", fields = {}) {
        this.url = url;
        this.method = method;
        this.fields = fields;
    }
    setURL(url) {
        this.url = url.toString();
        return this;
    }
    setMethod(method) {
        this.method = method;
        return this;
    }
    setField(name, value) {
        this.fields[name] = value;
        return this;
    }
    setFields(assignments) {
        Object.assign(this.fields, assignments);
        return this;
    }
    buildHtml(fullPage = false) {
        const fields = Object.entries(this.fields).map(([key, value]) => `<input type="hidden" name="${key}" value="${value}" />`);
        const id = `form${uniqueId()}`;
        const form = `<form method="${this.method}" action="${this.url}" id="${id}">
    ${fields.join("\n")}
</form>

<script>(function(){ document.getElementById('${id}').submit(); })();</script>`;
        return fullPage
            ? `<!DOCTYPE html>
<html>
<head>  
  <meta charset="utf-8" />
  <meta name="robots" content="noindex, nofollow, noarchive" />
  <title>Redirection...</title>
</head>  
<body>${form}</body>
</html>`
            : form;
    }
}
exports.FormBuilder = FormBuilder;
