// yoomoney-sdk/dist/cjs/fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const axios = require("axios"),
  querystring = require("querystring");
const _interopDefaultLegacy = (e) =>
  e && typeof e === "object" && "default" in e ? e["default"] : e;
const axios__default = /*#__PURE__*/ _interopDefaultLegacy(axios);
/**
 *
 * @param {string} url URL адрес запроса
 * @param {QueryStringifiable} parameters Параметры запроса
 * @param {Record<string, string>=} [headers={}] Заголовки запроса
 * @param {Agent} [agent] Агент запроса
 *
 * @return {Promise<any>} JSON в ответе
 */
async function fetch(url, parameters, headers = {}, agent) {
  return await axios__default
    .post(url, querystring.stringify(parameters), {
      headers: {
        "User-Agent": "yoomoney-sdk/1.5.5 (+https://npmjs.com/package/yoomoney-sdk)",
        ...headers,
        "Content-Type": "application/x-www-form-urlencoded",
        Accept: "application/json"
      },
      httpAgent: agent,
      httpsAgent: agent,
      responseType: "json"
    })
    .then((response) => response.data)
    .catch((error) =>
      axios__default.isAxiosError(error) && error.response
        ? error.response.data
        : Promise.reject(error)
    );
}
exports.fetch = fetch;
